core/services/ocr2/plugins/ocr2keeper/evm21/upkeep_state_store.go

```diff
@@ -1,97 +1,95 @@
 package evm
 
 import (
-	"math/big"
+	"context"
 	"sync"
 
 	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg/v3/types"
 
-	mapset "github.com/deckarep/golang-set/v2"
-
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
 )
 
-const BlockKeySeparator = "|"
-
 type upkeepState struct {
-	payload  *ocr2keepers.UpkeepPayload
-	state    *ocr2keepers.UpkeepState
-	block    int64
-	upkeepId string
+	upkeepID ocr2keepers.UpkeepIdentifier
+	workID   string
+	state    ocr2keepers.UpkeepState
+	block    uint64
 }
 
 type UpkeepStateReader interface {
-	// SelectByUpkeepIDsAndBlockRange retrieves upkeep states for provided upkeep ids and block range, the result is currently not in particular order
-	SelectByUpkeepIDsAndBlockRange(upkeepIds []*big.Int, start, end int64) ([]*ocr2keepers.UpkeepPayload, []*ocr2keepers.UpkeepState, error)
+	SelectByWorkIDs(workIDs ...string) (map[string]ocr2keepers.UpkeepState, error)
 }
 
-type UpkeepStateUpdater interface {
-	SetUpkeepState(ocr2keepers.UpkeepPayload, ocr2keepers.UpkeepState) error
-}
+var (
+	_ ocr2keepers.UpkeepStateUpdater = &UpkeepStateStore{}
+	_ UpkeepStateReader              = &UpkeepStateStore{}
+)
 
 type UpkeepStateStore struct {
-	mu         sync.RWMutex
-	statesByID map[string]*upkeepState
-	states     []*upkeepState
-	lggr       logger.Logger
+	mu        sync.RWMutex
+	workIDIdx map[string]*upkeepState
+	states    []*upkeepState
```

Comment:
a single map should be sufficient

Reply:
Keep in mind this is very temporary. I'll be doing a fast follower with the actual database implementation.

Reply:
done